File: Source/cmdApp/method/SignalInterval.m
classdef SignalInterval < handle
    % SignalInterval 类
    % 表示一个信号区间，包括信号数据、采样参数、频谱分析结果等

    properties (GetAccess = public, SetAccess = public)
       signal            (1, :) double = [-1 1]  % 该区间的信号数据
       samplingTime      (1, 1) double = 1       % 该信号的采样时间（秒）
       samplingFrequency (1, 1) double = 2       % 该信号的采样频率
       frequency         (1, :) double           % FFT 后的频率轴
       amplitude         (1, :) double           % FFT 后的幅值
       dominantFrequency (1, :) double         % 主频率成分（幅值最大者）
       dominantAmplitude (1, :) double         % 主频率对应的幅值
    end

    properties (Access = private)
        frequencyList    (1, :) double = 1      % 频率成分列表
        nSignal          (1, 1) double = 2      % 信号长度
    end
	
	

    methods (Access = public)
        % ===== 构造函数：根据频率列表生成信号区间 =====
        function si = SignalInterval(frequency_list, n_interpolation, sampling_time)
            % 内部创建 PseudoRandomSignal 对象并生成信号  --PseudoRandomSignal 
            prs = PseudoRandomSignal(frequency_list);
            si.signal            = prs.Sampling(n_interpolation, sampling_time);
            si.frequencyList     = frequency_list;
            si.samplingTime      = sampling_time;
            si.samplingFrequency = n_interpolation * prs.nSequence;
            si.nSignal           = length(si.signal);
            delete(prs);  % 用完即释放

            % 计算该信号的频谱信息
            si.Spectrum();
            si.MarkSpectrum();
        end



        % ===== 重采样（插值）方法 =====
        function si = Sampling(si, n_interpolation)
            arguments
                si                    SignalInterval
                n_interpolation (1,1) double = 1
            end
            % 提高信号的采样密度（通过重复信号点）
            si.samplingFrequency = n_interpolation * si.samplingFrequency; 
            si.signal            = repelem(si.signal, n_interpolation);
            si.nSignal           = length(si.signal);
            si.Spectrum();       % 重新计算频谱
            si.MarkSpectrum();   % 重新标记主频率
        end
    end



    methods (Access = private)
        % ===== 计算频谱（FFT）=====
        function si = Spectrum(si)
            n = si.nSignal /2;
            a = fft(si.signal); % 快速傅里叶变换
            si.frequency = (0:n-1)/si.samplingTime;  % 频率轴
            si.amplitude = abs(a(1:n))/n;  % 幅值（归一化）
        end

        % ===== 标记主频率（幅值最大的几个频率）=====
        function si = MarkSpectrum(si)
            [~, index] = sort(si.amplitude,'descend');  % 按幅值降序排序
            index = index(1:length(si.frequencyList));  % 取前 N 个（N = 输入频率个数）
            si.dominantFrequency = si.frequency(index);
            si.dominantAmplitude = si.amplitude(index);
        end
		
		
		
    end
end

File: Source/cmdApp/SignalInterval.cs
// SignalInterval.cs
using System;
using System.Linq;
using MathNet.Numerics.IntegralTransforms;
namespace cmdApp
{
    public class SignalInterval
    {
        public int[] Signal { get; private set; }
        public double SamplingTime { get; private set; }
        public double SamplingFrequency { get; private set; }
        public double[] Frequency { get; private set; }         // FFT 频率轴（只取前一半）
        public double[] Amplitude { get; private set; }        // FFT 幅值
        public double[] DominantFrequency { get; private set; } // 主频率（数组）
        public double[] DominantAmplitude { get; private set; } // 主频率对应幅值

        public double[] FrequencyList { get; set; }

        public SignalInterval(double[] frequencyList, int nInterpolation, double samplingTime)
        {
            FrequencyList = frequencyList;

            var prs = new PseudoRandomSignal(frequencyList);
            Signal = prs.Sampling(nInterpolation, (int)samplingTime);

            this.SamplingTime = samplingTime;
            SamplingFrequency = nInterpolation * prs.NSequence;

            ComputeSpectrum();
        }

        private void ComputeSpectrum()
        {
            int n = Signal.Length;
            double[] signalArray = Signal.Select(x => (double)x).ToArray();

            var complexSignal = signalArray.Select(x => new System.Numerics.Complex(x, 0)).ToArray();
            Fourier.Forward(complexSignal, FourierOptions.Matlab);

            int halfN = n / 2;
            Frequency = new double[halfN];
            Amplitude = new double[halfN];

            for (int i = 0; i < halfN; i++)
            {
                Frequency[i] = i / SamplingTime;                  // 频率 = idx / 时间
                Amplitude[i] = complexSignal[i].Magnitude / halfN; // 幅值归一化
            }

            // 找出幅值最大的前 N 个频率，N = 输入频率个数
            int topN = FrequencyList.Length;
            var sorted = Amplitude
                .Select((val, idx) => new { Value = val, Index = idx })
                .OrderByDescending(x => x.Value)
                .Take(topN)
                .ToArray();

            DominantFrequency = sorted.Select(x => Frequency[x.Index]).ToArray();
            DominantAmplitude = sorted.Select(x => Amplitude[x.Index]).ToArray();
        }
    }
}